In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

## 1. Introduction
--------------------

### What does a Null value look like? 
- Shows up as 'NaN'

### Why is it important to address?
- Models can't take NaN as input

### Types of Null values?
- Numerical: Eg - Number of rooms in the house
- Categorical: Eg - Name of housing society

In [2]:
#df = pd.read_csv('/kaggle/input/bengaluru-house-price-data/Bengaluru_House_Data.csv')
df = pd.read_csv('Bengaluru_House_Data.csv')
df

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK      NaN       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK      NaN       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK      NaN       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

**Check for NULL values**

In [3]:
# Check for NULL values
df.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

## 2. **Address NULL values**
-------------------------------
Using Pandas, we can either drop that row, or fill it
> *Filling of values is called as Imputation*


Let's look at our dataset in more detail: 

- Location: Just one value -> Drop

- Size: Numerical -> 16 -> Drop

- Society: Categorical -> 5502 -> Fill

- Bath: Numerical -> 73 -> Fill

- Balcony: Numerical -> 609 -> Fill



### 2a. **Fill All Columns with common value**

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html

> **Fill all Null Values in the dataset with a single value**

In [5]:
df2 = df.fillna(value = 0)
df2

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK        0       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK        0       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK        0       3600   5.0      0.0  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK        0        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

In [8]:
df2.isnull().sum()

area_type       0
availability    0
location        0
size            0
society         0
total_sqft      0
bath            0
balcony         0
price           0
dtype: int64

> **Propogate previous observed value**

In [9]:
df3 = df.fillna(method='ffill')
df3

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK  Theanmp       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK  Soiewre       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK  ArsiaEx       3600   5.0      0.0  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK  SollyCl        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

> **Propogate next valid observation to current**

In [10]:
df4 = df.fillna(method='bfill')
df4

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK  Soiewre       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK  DuenaTa       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK  Mahla T       3600   5.0      1.0  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

### 2b. **Fill Only Categorical Column: Society**


**Update the entire table**

Option 1:
> df = df.fillna(value=True)

Option 2:
> df.fillna(value=True, inplace=True)


**Update only one column**

Option 1:
> df = df.fillna(value={'column1':value})

Option 2: 
> df.fillna(value={'column1':value}, inplace=True)


**Update only multiple chosen columns**

Both options exist. You can mention in the dict, all the list of columns to be updated

> df.fillna(value={'column1':value, 'column2':value2}, inplace=True)


In [6]:
df5 = df.fillna(value={'society':'Unknown'})
df5

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK  Unknown       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK  Unknown       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK  Unknown       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK  Unknown        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

> **Fill numeric value**

In [7]:
# For a categorical value, we can use mean (can't calculate), and median (isn't a meaningful replacement)
# The most suited statistic is 'mode'

values = {
    'society' : df['society'].mode()[0]
}

df6 = df.fillna(value=values)
df6

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK  GrrvaGr       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK  GrrvaGr       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK  GrrvaGr       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK  GrrvaGr        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

In [13]:
df['society'].mode()

0    GrrvaGr
Name: society, dtype: object

In [14]:
df['society'].mode()[0]

'GrrvaGr'

> **bfill or ffill**

In [18]:
df['society'].fillna(method='bfill', inplace=True)
df


area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK  Soiewre       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK  DuenaTa       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK  Mahla T       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

### 2c. **Fill Multiple Numerical Columns: Bath, Balcony**


> **With Mean**

In [8]:
# Fill with mean/median
values = {
    'bath' : df['bath'].mean(),
    'balcony': df['balcony'].mean()
}

df7 = df6.fillna(value=values)
df7

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath   balcony   price  
0          2 BHK  Coomee        1056   2.0  1.000000   39.07  
1      4 Bedroom  Theanmp       2600   5.0  3.000000  120.00  
2          3 BHK  GrrvaGr       1440   2.0  3.000000   62.00  
3          3 BHK  Soiewre       1521   3.0  1.000000   95.00  
4          2 BHK  GrrvaGr       1200   2.0  1.000000   51.00  
...          ...      ...        ...   ...       ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0  0.000000  231.00  
13316      4 BHK  GrrvaGr       3600   5.0  1.584376  400.00  
13317      2 BHK  Mahla T       1141   2.0  1.000000   60.00  
13318      4 BHK  SollyCl       4689   4.0  1.000000  488.00  
13319      1 BHK  GrrvaGr        550   1.0  1.000000   17.00  

[13320 rows x 9 columns]

In [9]:
df['bath'].mean()

2.6926096474673513

In [24]:
df7.isnull().sum()

area_type        0
availability     0
location         1
size            16
society          0
total_sqft       0
bath             0
balcony          0
price            0
dtype: int64

> **With interpolate**

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html

In [10]:
df6

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK  GrrvaGr       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK  GrrvaGr       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK  GrrvaGr       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK  GrrvaGr        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

In [12]:
df6['balcony'] = df6['balcony'].interpolate(method='linear')
df6

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK  GrrvaGr       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK  GrrvaGr       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK  GrrvaGr       3600   5.0      0.5  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK  GrrvaGr        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

### 2d. **Drop Other chosen columns**

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html


In [31]:
df7

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath   balcony   price  
0          2 BHK  Coomee        1056   2.0  1.000000   39.07  
1      4 Bedroom  Theanmp       2600   5.0  3.000000  120.00  
2          3 BHK  GrrvaGr       1440   2.0  3.000000   62.00  
3          3 BHK  Soiewre       1521   3.0  1.000000   95.00  
4          2 BHK  GrrvaGr       1200   2.0  1.000000   51.00  
...          ...      ...        ...   ...       ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0  0.000000  231.00  
13316      4 BHK  GrrvaGr       3600   5.0  1.584376  400.00  
13317      2 BHK  Mahla T       1141   2.0  1.000000   60.00  
13318      4 BHK  SollyCl       4689   4.0  1.000000  488.00  
13319      1 BHK  GrrvaGr        550   1.0  1.000000   17.00  

[13320 rows x 9 columns]

In [32]:
df7.isnull().sum()

area_type        0
availability     0
location         1
size            16
society          0
total_sqft       0
bath             0
balcony          0
price            0
dtype: int64

In [13]:
df9 = df7.dropna(subset=['size', 'location'])
df9

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath   balcony   price  
0          2 BHK  Coomee        1056   2.0  1.000000   39.07  
1      4 Bedroom  Theanmp       2600   5.0  3.000000  120.00  
2          3 BHK  GrrvaGr       1440   2.0  3.000000   62.00  
3          3 BHK  Soiewre       1521   3.0  1.000000   95.00  
4          2 BHK  GrrvaGr       1200   2.0  1.000000   51.00  
...          ...      ...        ...   ...       ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0  0.000000  231.00  
13316      4 BHK  GrrvaGr       3600   5.0  1.584376  400.00  
13317      2 BHK  Mahla T       1141   2.0  1.000000   60.00  
13318      4 BHK  SollyCl       4689   4.0  1.000000  488.00  
13319      1 BHK  GrrvaGr        550   1.0  1.000000   17.00  

[13303 rows x 9 columns]

In [14]:
df9.isnull().sum()

area_type       0
availability    0
location        0
size            0
society         0
total_sqft      0
bath            0
balcony         0
price           0
dtype: int64

# SUMMARY

1. The number of null rows determine if you have to drop or fill. 

**FILL**

2. Categorical Variables: 
- If the data is time-series, or has some sequential dependency: Fill with bfill, ffill
- If you want to put one blanket value: can use 'Mode', or a constant 'Unknown'

3. Numerical Variables:
- Can replace with single value
- Preferred: Replace with mean/median

4. Do a column-wise filling, rather than a common filling across the entire dataset (which might be quicker - good for v1, but, less accurate). 

**DROP**

5. If the number is insignificant (< 5%), you can drop. You don't want to drop a large part of your dataset. 
